exchanges/subscription/template_test.go

```diff
@@ -0,0 +1,99 @@
+package subscription
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/kline"
+)
+
+// TestExpandTemplates exercises ExpandTemplates
+func TestExpandTemplates(t *testing.T) {
+	t.Parallel()
+
+	e := &mockEx{
+		tpl: "subscriptions.tmpl",
+	}
+
+	// Functionality tests
+	l := List{
+		{Channel: "feature1"},
+		{Channel: "feature2", Asset: asset.All, Pairs: currency.Pairs{btcusdtPair, ethusdcPair}, Interval: kline.FifteenMin},
+		{Channel: "feature3", Asset: asset.All, Pairs: currency.Pairs{btcusdtPair, ethusdcPair}, Levels: 100},
+		{Channel: "feature4", Authenticated: true},
+		{Channel: "feature1", QualifiedChannel: "just one sub already processed"},
+	}
+	got, err := l.ExpandTemplates(e)
+	require.NoError(t, err, "ExpandTemplates must not error")
+	exp := List{
+		{Channel: "feature1", QualifiedChannel: "feature1"},
+		{Channel: "feature2", QualifiedChannel: "spot-feature2@15m", Asset: asset.Spot, Pairs: currency.Pairs{btcusdtPair, ethusdcPair}, Interval: kline.FifteenMin},
+		{Channel: "feature2", QualifiedChannel: "future-feature2@15m", Asset: asset.Futures, Pairs: currency.Pairs{btcusdtPair, ethusdcPair}, Interval: kline.FifteenMin},
+		{Channel: "feature3", QualifiedChannel: "spot-USDTBTC-feature3@100", Asset: asset.Spot, Pairs: currency.Pairs{btcusdtPair}, Levels: 100},
+		{Channel: "feature3", QualifiedChannel: "spot-USDCETH-feature3@100", Asset: asset.Spot, Pairs: currency.Pairs{ethusdcPair}, Levels: 100},
+		{Channel: "feature3", QualifiedChannel: "future-USDTBTC-feature3@100", Asset: asset.Futures, Pairs: currency.Pairs{btcusdtPair}, Levels: 100},
+		{Channel: "feature3", QualifiedChannel: "future-USDCETH-feature3@100", Asset: asset.Futures, Pairs: currency.Pairs{ethusdcPair}, Levels: 100},
+		{Channel: "feature1", QualifiedChannel: "just one sub already processed"},
+	}
+
+	if !equalLists(t, exp, got) {
+		t.FailNow() // If the first list isn't equal testing it again will duplicate test failures
+	}
+
+	e.auth = true
+	got, err = l.ExpandTemplates(e)
+	require.NoError(t, err, "ExpandTemplates must not error")
+	exp = append(exp,
+		&Subscription{Channel: "feature4", QualifiedChannel: "feature4-authed"},
+	)
+	equalLists(t, exp, got)
+
+	_, err = List{{Channel: "feature2", Asset: asset.Spot}}.ExpandTemplates(e)
+	assert.ErrorIs(t, err, errAssetTemplateWithoutAll, "Should error correctly on xpand Assets without All")
```

Comment:
```suggestion
	assert.ErrorIs(t, err, errAssetTemplateWithoutAll, "Should error correctly on xpand assets without All")
```
Noticing below that pairs and assets is lowercase but assets is capitalised here perhaps to give emphasis on the type you're describing, but these should all be lowercase IMO.

Reply:
Yep, one or the other, but consistent.

Fixed 315b1c0